### ISSUE TITLE
Swagger Generator

### ISSUE THREAD

**Issue by amcdnl** [pleerock/routing-controllers#11]
Propose implementing swagger generation for route definitions as discussed on [swagger-ts](https://github.com/lukeautry/swagger-ts/issues/2) repo. I've talked to @lukeautry and he is interested in collabing on this.

**References**
- [class validator issue](https://github.com/pleerock/class-validator/issues/5)

**Comment by awaragi** [comment]
In case anyone wants to provide some feedback on my development of this feature https://github.com/pleerock/routing-controllers/pull/82

**Comment by andreymaznyak** [comment]
[I have a solution based on tsoa](https://github.com/andreymaznyak/tsoa)

**Comment by amcdnl** [comment]
@andreymaznyak - would u be interested in collaborating on this? I'm working on a project and I need it to have first class swagger support ( https://github.com/swimlane/trafficlight ).

**Comment by andreymaznyak** [comment]
@amcdnl Sure, i created pull request in your project ( https://github.com/swimlane/trafficlight/pull/5 )

**Comment by epiphone** [comment]
Hi, I made a stab at it [here](https://github.com/epiphone/class-validator-jsonschema). Would be glad to hear any feedback!

**Comment by jasperblues** [comment]
Look forward to trying @epiphone 
suggestion: perhaps the Readme for _this_ library could link to supporting libraries?

**Comment by NoNameProvided** [comment]
@jasperblues Yes, we totally can. 

I had no time to check these libs out, what is the status of these currently? We can link libraries in the README, but they must work for that.

**Comment by pairmix** [comment]
Hi there,
What about this feature?
Is https://github.com/epiphone/routing-controllers-openapi the current project?

**Comment by Marcus-Rise** [comment]
I'm interested in this feature too

**Comment by feimosi** [comment]
[routing-controllers-openapi](https://github.com/epiphone/routing-controllers-openapi) works really well for us.

**Comment by leonardlin** [comment]
https://github.com/epiphone/routing-controllers-openapi has been updated to match the updated routing-controllers v0.10.0

**Comment by leonardlin** [comment]
isn't this feature provided through https://github.com/epiphone/routing-controllers-openapi?

**Comment by douglasg14b** [comment]
Yeah, but it appears to be unmaintained, which means it's 1 breaking change away from dead

**Comment by e22m4u** [comment]
I made an example with swagger-ui.
It may be useful, but it requires to write openapi schema right in decorators.
https://github.com/e22m4u/routing-controllers-swagger-example

**Comment by e22m4u** [comment]
Swagger integration example
https://github.com/e22m4u/routing-controllers-swagger-example

```ts
const ROLE_SCHEMA: OASchemaObject = {
  type: OADataType.OBJECT,
  properties: {
    id: {type: OADataType.NUMBER},
    name: {type: OADataType.STRING},
  }
};

@oaTag()
@Controller()
export class RoleController {
  @oaOperation({
    method: OAOperationMethod.GET,
    path: '/roles',
  })
  @oaResponse({
    statusCode: 200,
    mediaType: OAMediaType.APPLICATION_JSON,
    description: 'Response example',
    schema: {
      type: OADataType.ARRAY,
      items: ROLE_SCHEMA,
    },
  })
  @Get('/roles')
  find() {
    return [
      {id: 1, name: 'admin'},
      {id: 2, name: 'manager'},
      {id: 3, name: 'user'},
    ];
  }
}
```
